File: Assets/Script/Lib/Agent/Tool/Batch.ts
// @preview-file off clear
import type { AgentToolName } from 'Agent/Tool/Types';

export interface AgentToolBatchItem {
	tool: AgentToolName;
	toolCallId: string;
	params: Record<string, unknown>;
}

export interface AgentToolDecisionItem {
	tool: AgentToolName;
	toolCallId?: string;
	params: Record<string, unknown>;
}

export interface AgentToolBatch<T extends AgentToolBatchItem> {
	isConcurrencySafe: boolean;
	actions: T[];
}

export function cloneAgentToolParams(value: Record<string, unknown>): Record<string, unknown> {
	function clone(item: unknown): unknown {
		if (item === undefined) return item;
		if (Array.isArray(item)) return item.map(child => clone(child));
		if (typeof item === "object") {
			const output: Record<string, unknown> = {};
			for (const key in item as Record<string, unknown>) output[key] = clone((item as Record<string, unknown>)[key]);
			return output;
		}
		return item;
	}
	return clone(value) as Record<string, unknown>;
}

export function areAgentToolParamsEqual(left: unknown, right: unknown): boolean {
	if (left === right) return true;
	if (left === undefined || right === undefined) return false;
	if (Array.isArray(left) || Array.isArray(right)) {
		if (!Array.isArray(left) || !Array.isArray(right) || left.length !== right.length) return false;
		for (let i = 0; i < left.length; i++) {
			if (!areAgentToolParamsEqual(left[i], right[i])) return false;
		}
		return true;
	}
	if (typeof left === "object" && typeof right === "object") {
		let leftCount = 0;
		for (const key in left as Record<string, unknown>) {
			leftCount++;
			if (!areAgentToolParamsEqual((left as Record<string, unknown>)[key], (right as Record<string, unknown>)[key])) return false;
		}
		let rightCount = 0;
		for (const _key in right as Record<string, unknown>) rightCount++;
		return leftCount === rightCount;
	}
	return false;
}

function getReadBatchItems(params: Record<string, unknown>): Record<string, unknown>[] | undefined {
	if (!Array.isArray(params.reads)) return undefined;
	return params.reads.map(item => cloneAgentToolParams(item as Record<string, unknown>));
}

function getBuildBatchPaths(params: Record<string, unknown>): string[] | undefined {
	return Array.isArray(params.paths) ? (params.paths as string[]).slice() : undefined;
}

/**
 * Normalize consecutive compatible calls from one model response into the
 * existing batch form. The first call id is retained because history records
 * the normalized assistant message, while the raw provider response remains
 * available in the per-step debug output.
 */
export function coalesceCompatibleAgentToolCalls<T extends AgentToolDecisionItem>(actions: T[]): T[] {
	const output: T[] = [];
	let index = 0;
	while (index < actions.length) {
		const first = actions[index];
		if (first.tool !== "read_file" && first.tool !== "build") {
			output.push(first);
			index++;
			continue;
		}
		let end = index;
		while (end + 1 < actions.length && actions[end + 1].tool === first.tool) end++;
		if (end === index) {
			output.push(first);
			index++;
			continue;
		}
		if (first.tool === "read_file") {
			const reads: Record<string, unknown>[] = [];
			let compatible = true;
			for (let i = index; i <= end; i++) {
				const items = getReadBatchItems(actions[i].params);
				if (items === undefined) { compatible = false; break; }
				reads.push(...items);
			}
			if (compatible) {
				output.push({ ...first, params: { reads } });
				index = end + 1;
				continue;
			}
		} else {
			const paths: string[] = [];
			let compatible = true;
			for (let i = index; i <= end; i++) {
				const items = getBuildBatchPaths(actions[i].params);
				if (items === undefined) { compatible = false; break; }
				paths.push(...items);
			}
			if (compatible) {
				output.push({ ...first, params: { paths } });
				index = end + 1;
				continue;
			}
		}
		for (let i = index; i <= end; i++) output.push(actions[i]);
		index = end + 1;
	}
	return output;
}

export function partitionAgentToolCalls<T extends AgentToolBatchItem>(
	actions: T[],
	isParallelSafe: (this: void, tool: AgentToolName) => boolean
): AgentToolBatch<T>[] {
	const batches: AgentToolBatch<T>[] = [];
	for (let i = 0; i < actions.length; i++) {
		const action = actions[i];
		const isSafe = isParallelSafe(action.tool);
		const last = batches.length > 0 ? batches[batches.length - 1] : undefined;
		if (isSafe && last?.isConcurrencySafe === true) last.actions.push(action);
		else batches.push({ isConcurrencySafe: isSafe, actions: [action] });
	}
	return batches;
}

File: Assets/Script/Lib/Agent/Tool/Batch.lua
-- [ts]: Batch.ts
local ____lualib = require("lualib_bundle") -- 1
local __TS__ArrayIsArray = ____lualib.__TS__ArrayIsArray -- 1
local __TS__ArrayMap = ____lualib.__TS__ArrayMap -- 1
local ____exports = {} -- 1
function ____exports.cloneAgentToolParams(value) -- 15
	local function clone(item) -- 16
		if item == nil then -- 16
			return item -- 17
		end -- 17
		if __TS__ArrayIsArray(item) then -- 17
			return __TS__ArrayMap( -- 18
				item, -- 18
				function(____, child) return clone(child) end -- 18
			) -- 18
		end -- 18
		if type(item) == "table" then -- 18
			local output = {} -- 20
			for key in pairs(item) do -- 21
				output[key] = clone(item[key]) -- 21
			end -- 21
			return output -- 22
		end -- 22
		return item -- 24
	end -- 16
	return clone(value) -- 26
end -- 15
function ____exports.areAgentToolParamsEqual(left, right) -- 29
	if left == right then -- 29
		return true -- 30
	end -- 30
	if left == nil or right == nil then -- 30
		return false -- 31
	end -- 31
	if __TS__ArrayIsArray(left) or __TS__ArrayIsArray(right) then -- 31
		if not __TS__ArrayIsArray(left) or not __TS__ArrayIsArray(right) or #left ~= #right then -- 31
			return false -- 33
		end -- 33
		do -- 33
			local i = 0 -- 34
			while i < #left do -- 34
				if not ____exports.areAgentToolParamsEqual(left[i + 1], right[i + 1]) then -- 34
					return false -- 35
				end -- 35
				i = i + 1 -- 34
			end -- 34
		end -- 34
		return true -- 37
	end -- 37
	if type(left) == "table" and type(right) == "table" then -- 37
		local leftCount = 0 -- 40
		for key in pairs(left) do -- 41
			leftCount = leftCount + 1 -- 42
			if not ____exports.areAgentToolParamsEqual(left[key], right[key]) then -- 42
				return false -- 43
			end -- 43
		end -- 43
		local rightCount = 0 -- 45
		for _key in pairs(right) do -- 46
			rightCount = rightCount + 1 -- 46
		end -- 46
		return leftCount == rightCount -- 47
	end -- 47
	return false -- 49
end -- 29
function ____exports.partitionAgentToolCalls(actions, isParallelSafe) -- 52
	local batches = {} -- 56
	do -- 56
		local i = 0 -- 57
		while i < #actions do -- 57
			local action = actions[i + 1] -- 58
			local isSafe = isParallelSafe(action.tool) -- 59
			local last = #batches > 0 and batches[#batches] or nil -- 60
			if isSafe and (last and last.isConcurrencySafe) == true then -- 60
				local ____last_actions_2 = last.actions -- 60
				____last_actions_2[#____last_actions_2 + 1] = action -- 61
			else -- 61
				batches[#batches + 1] = {isConcurrencySafe = isSafe, actions = {action}} -- 62
			end -- 62
			i = i + 1 -- 57
		end -- 57
	end -- 57
	return batches -- 64
end -- 52
return ____exports -- 52